packages/graphiql/src/state/GraphiQLSessionProvider.tsx

```diff
@@ -0,0 +1,223 @@
+  dispatch: () => null,
+  editorLoaded: () => null,
+  ...initialState,
+};
+
+export const SessionContext = React.createContext<
+  SessionState & SessionHandlers
+>(initialContextState);
+
+export const useSessionContext = (): SessionState & SessionHandlers =>
+  React.useContext(SessionContext);
+
+export function getSessionReducer(schema: GraphQLSchema) {
+  return (state: SessionState, { type: actionType, payload }: ActionState) => {
+    switch (actionType) {
+      case actionTypes.editor_loaded: {
+        const { context, editor } = payload as EditorPayload;
+        state.editors[context as EditorContexts] = editor as CodeMirror.Editor;
+        return state;
+      }
+      case actionTypes.operation_changed: {
+        const { value } = payload as ChangeValuePayload;
+        state.operation.text = value as string;
+        return {
+          ...state,
+          ...getQueryFacts(schema as GraphQLSchema, value),
+        };
+      }
+      case actionTypes.variables_changed: {
+        const { value } = payload as ChangeValuePayload;
+        state.variables.text = value as string;
+        return state;
+      }
+      case actionTypes.operation_succeeded: {
+        const { result } = payload as SuccessPayload;
+        state.results.text = result as string;
+        state.results.formattedText = JSON.stringify(result, null, 2);
+        state.operationErrors = null;
+        return state;
+      }
+      case actionTypes.operation_errored: {
+        const { error } = payload as ErrorPayload;
+        state.operationErrors = [error.toString()];
+        return state;
+      }
+      default: {
+        return state;
+      }
+    }
+  };
+}
+
+export type SessionProviderProps = {
+  sessionId: number;
+  fetcher?: Fetcher;
+  session?: SessionState;
+  children: any;
+};
+
+export function SessionProvider({
+  sessionId,
+  fetcher = defaultFetcher,
+  session,
+  ...props
+}: SessionProviderProps) {
+  const schemaState = React.useContext(SchemaContext);
+
+  const [state, dispatch] = useReducers<SessionState, AT, ActionState>({
+    reducers: [getSessionReducer(schemaState.schema as GraphQLSchema)],
+    init: () => initialState,
+  });
+
+  const operationError = (error: Error) =>
+    dispatch({
+      type: actionTypes.operation_errored,
+      payload: { error, sessionId },
+    });
+  const editorLoaded = (context: EditorContexts, editor: CodeMirror.Editor) => {
+    dispatch({
+      type: actionTypes.operation_errored,
+      payload: { editor, context },
+    });
+  };
+  const changeOperation = (operationText: string) => {
+    dispatch({
+      type: actionTypes.operation_changed,
+      payload: { value: operationText, sessionId },
+    });
+  };
+
+  const changeVariables = (variablesText: string) =>
+    dispatch({
+      type: actionTypes.variables_changed,
+      payload: { value: variablesText, sessionId },
+    });
+  const executeOperation = async (
+    sessionState: SessionState,
```

Comment:
if feels kind of strange that you have to pass the session state back into this function. Shouldn't the outer scope provide everything this function needs or are you trying to just encapsulate it all? lines like this one just feel strange to me:

https://github.com/graphql/graphiql/pull/1359/files#diff-de4c1a6e083a392d7c4acd996d95106dR49